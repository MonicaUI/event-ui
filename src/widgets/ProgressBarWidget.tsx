import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import withForm from 'nucleus-form/src/components/withForm';
import ProgressBarWidget from 'event-widgets/lib/ProgressBar/ProgressBarWidget';
import {
  saveRegistrationAndRouteToPage,
  withScrollToFirstError
} from './RegistrationNavigator/RegistrationNavigatorWidget';
import { getRegistrationPathIdOrDefault } from '../redux/selectors/currentRegistrationPath';
import { isRegistrationPage } from '../redux/website/registrationProcesses';
import { getCurrentPageId } from '../redux/pathInfo';
import { getIn } from 'icepick';
import { nucleusFormShape } from 'nucleus-form/src/NucleusForm';
import { get } from 'lodash';

const MAX_PROGRESS = 100;
const DEFAULT_PROGRESS = 25;
const CURRENT_PAGE = 'X';
const TOTAL_PAGE = 'Y';

const withContextNucleusForm = () => {
  return () => WrappedComponent => {
    return class WithContextNucleusForm extends React.Component {
      static displayName = 'WithFormProgressBarWidget';
      static contextTypes = {
        nucleusForm: nucleusFormShape
      };

      hasNucleusFormInContext: $TSFixMe;

      constructor(props, context) {
        super(props, context);
        this.hasNucleusFormInContext = Boolean(context.nucleusForm);
      }

      render() {
        if (this.hasNucleusFormInContext) {
          // eslint-disable-next-line react/prop-types
          return <WrappedComponent {...this.props}>{this.props.children}</WrappedComponent>;
        }
        return null;
      }
    };
  };
};

export function getPageNames(state: $TSFixMe, pageIds: $TSFixMe): $TSFixMe {
  const userText = state?.localizedUserText?.currentLocale
    ? get(state.localizedUserText.localizations, state.localizedUserText.currentLocale, null)
    : null;
  return (pageIds || []).map(
    pageId =>
      state.website.pages[pageId] &&
      state.text.translate(get(userText, 'website.pages.' + pageId + '.name', state.website.pages[pageId].name))
  );
}

export default withContextNucleusForm()(
  // @ts-expect-error ts-migrate(2554) FIXME: Expected 0 arguments, but got 1.
  withForm(
    () => ({}),
    formActions => ({ submitForm: formActions.submitForm })
  )
)(
  connect(
    (state: $TSFixMe) => {
      const currentPageId = getCurrentPageId(state);
      const isRegPage = isRegistrationPage(state, currentPageId);
      const currentRegPathId = isRegPage ? getRegistrationPathIdOrDefault(state) : '';
      const pageIds = getIn(state.website, [
        'pluginData',
        'registrationProcessNavigation',
        'registrationPaths',
        currentRegPathId,
        'pageIds'
      ]);
      const currentPageIndex = (pageIds || []).findIndex(id => id === currentPageId);

      return {
        pages: getPageNames(state, pageIds),
        pageIds,
        total: MAX_PROGRESS,
        digit:
          currentPageIndex !== -1 ? Math.ceil((currentPageIndex / pageIds.length) * MAX_PROGRESS) : DEFAULT_PROGRESS,
        currentPage: currentPageIndex !== -1 ? currentPageIndex + 1 : CURRENT_PAGE,
        totalPages: currentPageIndex !== -1 ? pageIds.length : TOTAL_PAGE,
        hideWidget: currentPageIndex === -1
      };
    },
    (dispatch: $TSFixMe, props: $TSFixMe) =>
      bindActionCreators(
        {
          onNavigateRequest: saveRegistrationAndRouteToPage.bind(null, withScrollToFirstError(props.submitForm))
        },
        dispatch
      )
  )(ProgressBarWidget)
);
