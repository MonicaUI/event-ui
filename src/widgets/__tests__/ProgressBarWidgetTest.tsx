import React from 'react';
import ProgressBarWidget from '../ProgressBarWidget';
import { getPageNames } from '../ProgressBarWidget';
import { shallow } from 'enzyme';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import withForm from 'nucleus-form/src/components/withForm';

jest.mock('../../redux/pathInfo', () => {
  return {
    getCurrentPageId: () => 'current-page'
  };
});

jest.mock('../../redux/website/registrationProcesses', () => {
  return {
    isRegistrationPage: () => true
  };
});

jest.mock('../../redux/selectors/currentRegistrant', () => {
  return {
    getRegistrationPathId: () => 'regpath1'
  };
});

jest.mock('../RegistrationNavigator/RegistrationNavigatorWidget', () => {
  return {
    saveRegistrationAndRouteToPage: () => true,
    withScrollToFirstError: () => true
  };
});

function getState() {
  const state = {
    text: {
      translate: x => x
    },
    website: ['pluginData', 'registrationProcessNavigation', 'registrationPaths', 'regpath1']
      .reverse()
      // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
      .reduce((prev, current) => ({ [current]: { ...prev } }), { pageIds: ['current-page', 'next-page'] })
  };
  (state.website as $TSFixMe).pages = {
    'current-page': {
      name: 'current-page'
    },
    'next-page': {
      name: 'next-page'
    }
  };

  return state;
}

async function dispatch(action) {
  if (typeof action === 'function') {
    await action(dispatch, getState);
  }
}

const subscribe = jest.fn();
const defaultProps = {
  classes: {},
  style: {},
  translate: c => c,
  store: { dispatch, getState, subscribe }
};

describe('ProgressBar produces props from state', () => {
  test('should match', () => {
    const widget = shallow(
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'withForm' does not exist on type 'JSX.In... Remove this comment to see the full error message
      <withForm>
        <ProgressBarWidget {...defaultProps} />
        {/* @ts-expect-error ts-migrate(2339) FIXME: Property 'withForm' does not exist on type 'JSX.In... Remove this comment to see the full error message */}
      </withForm>
    );
    expect(widget.props()).toMatchSnapshot();
  });
});

describe('getPageNames', () => {
  it('should get names of the pages for the progress bar when the pages has the correct pageIds', () => {
    const widgets = getPageNames(getState(), ['current-page', 'next-page']);
    expect(widgets.length).toBe(2);
  });
  it('should get names of the pages for the progress bar even when the pages do not have those pageIds', () => {
    const widgets = getPageNames(getState(), ['current-page1']);
    expect(widgets[0]).toBe(undefined);
  });
});
