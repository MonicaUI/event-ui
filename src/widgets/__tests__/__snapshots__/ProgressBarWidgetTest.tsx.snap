// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar produces props from state should match 1`] = `
Object {
  "children": <WithFormProgressBarWidget
    classes={Object {}}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [MockFunction],
      }
    }
    style={Object {}}
    translate={[Function]}
  />,
}
`;
